test/unit/mdc-data-table/foundation.test.js

```diff
@@ -0,0 +1,228 @@
+ * THE SOFTWARE.
+ */
+
+import td from 'testdouble';
+
+import {setupFoundationTest} from '../helpers/setup';
+import {verifyDefaultAdapter} from '../helpers/foundation';
+
+import {cssClasses, strings} from '../../../packages/mdc-data-table/constants';
+import {MDCDataTableFoundation} from '../../../packages/mdc-data-table/foundation';
+
+suite('MDCDataTableFoundation');
+
+test('default adapter returns a complete adapter implementation', () => {
+  verifyDefaultAdapter(MDCDataTableFoundation, [
+    'isRowsSelectable',
+    'registerHeaderRowCheckbox',
+    'registerRowCheckboxes',
+    'getRowElements',
+    'isHeaderRowCheckboxChecked',
+    'getRowCount',
+    'getSelectedRowCount',
+    'addClassAtRowIndex',
+    'removeClassAtRowIndex',
+    'setAttributeAtRowIndex',
+    'getAttributeAtRowIndex',
+    'getRowIndexByChildElement',
+    'setHeaderRowCheckboxIndeterminate',
+    'setHeaderRowCheckboxChecked',
+    'getRowIdAtIndex',
+    'setRowCheckboxCheckedAtIndex',
+    'notifyRowSelectionChanged',
+    'notifySelectedAll',
+    'notifyUnselectedAll',
+  ]);
+});
+
+/**
+ * @return {{mockAdapter: !MDCDataTableAdapter, foundation: !MDCDataTableFoundation}}
+ */
+function setupTest() {
+  const adapterFoundationPair = /** @type {{mockAdapter: !MDCDataTableAdapter, foundation: !MDCDataTableFoundation}} */
+    (setupFoundationTest(MDCDataTableFoundation));
+  adapterFoundationPair.foundation.init();
+  return adapterFoundationPair;
+}
+
+test('#layout should register header row checkbox only if table is selectable', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.isRowsSelectable()).thenReturn(true);
+  foundation.layout();
+  td.verify(mockAdapter.registerHeaderRowCheckbox(), {times: 1});
+  td.verify(mockAdapter.registerRowCheckboxes(), {times: 1});
+
+  td.reset();
+
+  td.when(mockAdapter.isRowsSelectable()).thenReturn(false);
+  foundation.layout();
+  td.verify(mockAdapter.registerHeaderRowCheckbox(), {times: 0});
+  td.verify(mockAdapter.registerRowCheckboxes(), {times: 0});
+});
+
+test('#layout should set header row checkbox checked when all row checkboxes are checked', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.getSelectedRowCount()).thenReturn(5);
+  td.when(mockAdapter.getRowCount()).thenReturn(5);
+
+  foundation.layout();
+  td.verify(mockAdapter.setHeaderRowCheckboxIndeterminate(false), {times: 1});
+  td.verify(mockAdapter.setHeaderRowCheckboxChecked(true), {times: 1});
+});
+
+test('#layout should set header row checkbox unchecked when all row checkboxes are unchecked', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.getSelectedRowCount()).thenReturn(0);
+  td.when(mockAdapter.getRowCount()).thenReturn(5);
+
+  foundation.layout();
+  td.verify(mockAdapter.setHeaderRowCheckboxIndeterminate(false), {times: 1});
+  td.verify(mockAdapter.setHeaderRowCheckboxChecked(false), {times: 1});
+});
```

Comment:
I think this is missing the indeterminate case?

Reply:
It was covered in another unit test case. But I added additional unit test case for `#layout` with indeterminate case.